In [233]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly as py
py.offline.init_notebook_mode(connected=True)
%matplotlib inline

##Dear Grader: IF you cannot see our graphs, please remember to pip all the libraries, including matplotlib, plotly and pyplot... 

In [234]:
data=pd.read_csv('breakfast_cereals.csv')

In [235]:
from sklearn import preprocessing
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
plt.rc("font", size=14)
sns.set(style="white")
sns.set(style="whitegrid", color_codes=True)

In [236]:
data.groupby('Manufacturer').mean()

Calories   Protein       Fat      Sodium     Fiber  \
Manufacturer                                                           
General Mills   106.590909  2.213636  1.309091  193.663636  1.177273   
Kelloggs        100.873913  2.486957  0.539130  163.678261  2.539130   
Nabisco          87.280000  2.880000  0.200000   29.000000  4.720000   
Post            102.911111  2.300000  0.822222  137.044444  2.500000   
Quaker Oats     108.571429  2.714286  1.714286  105.714286  1.285714   
Ralston Purina  115.714286  2.571429  1.142857  197.857143  2.000000   

                Carbohydrates     Sugar   Potassium   Vitamins  
Manufacturer                                                    
General Mills       14.172727  7.559091   79.522727  33.204545  
Kelloggs            14.152174  6.952174   93.734783  32.608696  
Nabisco             15.660000  2.200000  148.900000  10.000000  
Post                12.655556  8.166667  102.277778  23.755556  
Quaker Oats         14.857143  6.142857   78.571429  14.285714  
Ralston Purina      18.142857  5.857143  102.142857  25.000000

In [197]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

Realized accuracy isn't too good to predict amount of sugar level. Why not define healthiness with sodium level?
Sodium level (mg) >= 110 is unhealthy, else healthy. 

In [237]:
y = data['Sodium']*10
X = data.drop('Sodium', axis = 1).drop('Manufacturer', axis = 1).drop('Brand Name', axis =1)*10

In [250]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)
logreg = LogisticRegression()
logreg.fit(X_train, y_train)
pred_logreg = logreg.predict(X_test)/10
pred_logreg

/Users/oscarso/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/Users/oscarso/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:460: FutureWarning:

Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.



array([128. ,  95. , 140. , 128. , 125. ,   0. , 140. , 170. ,  95. ,
       140. , 140. , 157.9, 180. ,   0. , 128. , 240. , 140. , 140. ,
         0. ])

In [251]:
accuracy_score(y_true = y_test, y_pred = pred_logreg*10)

0.15789473684210525

In [253]:
count = 0
for i in y_test:
    if (i<110):
        if(pred_logreg[count]<110):
            print ("Healthy: Correct")
        else:
            print ("Healthy: Incorrect")
    else:
        if(pred_logreg[count]>=110):
            print("Unhealthy: Correct")
        else:
            print("Unhealthy: Incorrect")
    count = count + 1

Unhealthy: Correct
Unhealthy: Incorrect
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Incorrect
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Incorrect
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Healthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Healthy: Correct


In [254]:
accuracy = 16/19
print (accuracy)

0.8421052631578947


Looking at the visualizations, we see that sodium level is a pretty good indicator of whether a cereal is healthy or not. Thus, we decided to use a logistic regression model on the dataset except for non-numeric data and the feature itself to produce a guess-timate of the sodium level in each cereal. As above, we realized that predicting a value wasn't going to be very accurate, so we decided to provide a range that demonstrates whether a cereal is healthy or not. A cereal, in our case, is healthy if sodium levels are below 110mg and unhealthy if it is above or equal to 110mg. We trained and tested our dataset with 0.25 of the initially randomly arranged cereals data (not sorted like the visualization part). Comparing to the initial data, our logistic regression model predicts, with the generated sodium values, a cereal's healthiness correctly 84.2% of the time. Below, we will try to train all the data from our initial breakfast_cereals dataset and test its accuracy on a different dataset of more cereals. We will first do some preprocessing of the cereal.csv dataset to take the form of the original dataset so we can run the logistic regression on it. 

In [256]:
y = data['Sodium']*10
X = data.drop('Sodium', axis = 1).drop('Manufacturer', axis = 1).drop('Brand Name', axis =1)*10

In [258]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=1, random_state=0)
logreg = LogisticRegression()
logreg.fit(X_train, y_train)
pred_logreg = logreg.predict(X_test)/10
pred_logreg

/Users/oscarso/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.

/Users/oscarso/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:460: FutureWarning:

Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.



array([180.5])

In [260]:
data2=pd.read_csv('cereal.csv')

In [261]:
data2.head()

name mfr type  calories  protein  fat  sodium  fiber  \
0                  100% Bran   N    C        70        4    1     130   10.0   
1          100% Natural Bran   Q    C       120        3    5      15    2.0   
2                   All-Bran   K    C        70        4    1     260    9.0   
3  All-Bran with Extra Fiber   K    C        50        4    0     140   14.0   
4             Almond Delight   R    C       110        2    2     200    1.0   

   carbo  sugars  potass  vitamins  shelf  weight  cups     rating  
0    5.0       6     280        25      3     1.0  0.33  68.402973  
1    8.0       8     135         0      3     1.0  1.00  33.983679  
2    7.0       5     320        25      3     1.0  0.33  59.425505  
3    8.0       0     330        25      3     1.0  0.50  93.704912  
4   14.0       8      -1        25      3     1.0  0.75  34.384843

In [268]:
updated_data2 = data2.drop(columns = (['name','mfr','type','shelf','weight','cups','rating']))

In [269]:
updated_data2.head()

calories  protein  fat  sodium  fiber  carbo  sugars  potass  vitamins
0        70        4    1     130   10.0    5.0       6     280        25
1       120        3    5      15    2.0    8.0       8     135         0
2        70        4    1     260    9.0    7.0       5     320        25
3        50        4    0     140   14.0    8.0       0     330        25
4       110        2    2     200    1.0   14.0       8      -1        25

In [270]:
updated_data2.rename(index=str, columns={"calories": "Calories", "protein": "Protein", "fat": "Fat", "sodium": "Sodium",
                                         "fiber": "Fiber",
                                        "carbo": "Carbohydrates", "sugars": "Sugar", "potass": "Potassium", 
                                        "vitamins": "Vitamins"})


Calories  Protein  Fat  Sodium  Fiber  Carbohydrates  Sugar  Potassium  \
0         70        4    1     130   10.0            5.0      6        280   
1        120        3    5      15    2.0            8.0      8        135   
2         70        4    1     260    9.0            7.0      5        320   
3         50        4    0     140   14.0            8.0      0        330   
4        110        2    2     200    1.0           14.0      8         -1   
5        110        2    2     180    1.5           10.5     10         70   
6        110        2    0     125    1.0           11.0     14         30   
7        130        3    2     210    2.0           18.0      8        100   
8         90        2    1     200    4.0           15.0      6        125   
9         90        3    0     210    5.0           13.0      5        190   
10       120        1    2     220    0.0           12.0     12         35   
11       110        6    2     290    2.0           17.0      1        105   
12       120        1    3     210    0.0           13.0      9         45   
13       110        3    2     140    2.0           13.0      7        105   
14       110        1    1     180    0.0           12.0     13         55   
15       110        2    0     280    0.0           22.0      3         25   
16       100        2    0     290    1.0           21.0      2         35   
17       110        1    0      90    1.0           13.0     12         20   
18       110        1    1     180    0.0           12.0     13         65   
19       110        3    3     140    4.0           10.0      7        160   
20       100        3    0      80    1.0           21.0      0         -1   
21       110        2    0     220    1.0           21.0      3         30   
22       100        2    1     140    2.0           11.0     10        120   
23       100        2    0     190    1.0           18.0      5         80   
24       110        2    1     125    1.0           11.0     13         30   
25       110        1    0     200    1.0           14.0     11         25   
26       100        3    0       0    3.0           14.0      7        100   
27       120        3    2     160    5.0           12.0     10        200   
28       120        3    0     240    5.0           14.0     12        190   
29       110        1    1     135    0.0           13.0     12         25   
..       ...      ...  ...     ...    ...            ...    ...        ...   
47       100        2    1     220    2.0           15.0      6         90   
48       120        2    1     190    0.0           15.0      9         40   
49       140        3    2     220    3.0           21.0      7        130   
50        90        3    0     170    3.0           18.0      2         90   
51       130        3    2     170    1.5           13.5     10        120   
52       120        3    1     200    6.0           11.0     14        260   
53       100        3    0     320    1.0           20.0      3         45   
54        50        1    0       0    0.0           13.0      0         15   
55        50        2    0       0    1.0           10.0      0         50   
56       100        4    1     135    2.0           14.0      6        110   
57       100        5    2       0    2.7           -1.0     -1        110   
58       120        3    1     210    5.0           14.0     12        240   
59       100        3    2     140    2.5           10.5      8        140   
60        90        2    0       0    2.0           15.0      6        110   
61       110        1    0     240    0.0           23.0      2         30   
62       110        2    0     290    0.0           22.0      3         35   
63        80        2    0       0    3.0           16.0      0         95   
64        90        3    0       0    4.0           19.0      0        140   
65        90        3    0       0    3.0           20.0      0        120   
66       110        2    1

In [273]:
trial = updated_data2.drop(columns = (['sodium']))
pred_logreg = logreg.predict(trial*10)/10
pred_logreg

array([130. ,  15. , 260. , 140. ,  90. , 140. , 125. , 140. , 200. ,
         0. , 180. , 290. , 220. , 140. , 180. , 280. , 290. ,  90. ,
       180. , 140. ,  90. , 170. , 180. ,   0. , 125. , 220. ,   0. ,
       140. , 180.5, 200. ,  45. , 180. , 140. , 170. ,  75. , 220. ,
       180. , 180. , 200. , 200. , 290. , 150. , 180. , 136. , 140. ,
       140. , 100. , 140. , 190. , 140. ,   0. , 136. , 180.5, 320. ,
         0. ,   0. , 150. , 100. , 157.9, 140. ,   0. , 240. , 280. ,
         0. ,   0. ,   0. ,  70. , 230. , 200. , 200. , 126.7, 170. ,
       250. , 200. , 140. , 140. , 200. ])

In [275]:
count = 0
right = 0
for i in updated_data2['sodium']:
    if (i<110):
        if(pred_logreg[count]<110):
            print ("Healthy: Correct")
            right = right + 1
        else:
            print ("Healthy: Incorrect")
    else:
        if(pred_logreg[count]>=110):
            print("Unhealthy: Correct")
            right = right + 1
        else:
            print("Unhealthy: Incorrect")
    count = count + 1

Unhealthy: Correct
Healthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Incorrect
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Incorrect
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Healthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Healthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Incorrect
Unhealthy: Correct
Unhealthy: Correct
Healthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Healthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Healthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Healthy: Incorrect
Healthy: Incorrect
Unhealthy: Correct
Unhealthy: Incorrect
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Correct
Unhealthy: Incorrect
Unhealthy: Correct
Unhealthy: Cor

In [276]:
accuracy = right / count
print(accuracy)

0.8961038961038961


From using the trained data from our breakfast_cerealsl.csv dataset, we are successfully able to predict, with high confidence, the healthiness or unhealthiness of a cereal based off predicting sodium levels from other factors in the cereal even if a new dataset/new data is introduced. 